hips/hip-9999.md

```diff
@@ -0,0 +1,81 @@
+---
+hip: 9999
+title: "Hook Parallelism"
+authors: [ "Andrew Baehre <abaehre@morningconsult.com>" ]
+created: "2020-12-02"
+type: "feature"
+status: "draft"
+---
+
+## Abstract
+
+Moving from Helm 2 to Helm 3 the ability to run hooks in parallel for tests was
+lost. This feature would allow for hooks to be run in parallel instead of
+serially for not only tests, but for hooks in general.
+
+## Motivation
+
+Currently, there is no way to run hooks in parallel. While generally this would
+be fine, for running tests in a CI environment the amount of time it takes to
+run several tests is too long. By running in parallel running helm tests in CI
+becomes much more tenable.
+
+## Rationale
+
+The proposed design is to add a flag called "hook-parallelism", which
```

Comment:
When designing proposals, we commonly refer to the [User Profile](https://github.com/helm/community/blob/master/user-profiles.md) this feature is intended to target.

If I'm reading this proposal correctly, this proposal is aimed at the Application Operator. In other words, it written for the person who is running `helm test`, not the person who wrote the tests themselves (the Application Distributor).

With this design, the Application Distributor does not have any control whether the tests should be run in serial or in parallel. That seems to me like a major design flaw with this proposal. It is common for someone to write a test that cannot be run in parallel or vice-versa (i.e. testing a worklflow, or testing a race condition).

I would highly re-consider who is the intended target audience for this proposal and how it fits into the overall design. i.e. how the Application Distributor can inform the Application Operator what tests can be run in parallel, or perhaps not at all. As in, the Application Distributor should have full control over which tests can be run in parallel vs what cannot.

For example, have a look at Go's `testing` package and see how they've designed test parallelism into the package. The person running `go test` has no control over which tests can be run in parallel; they can only control which tests/subtests are run.

https://golang.org/pkg/testing/#T.Parallel